### PR TITLE
PC-1463: Improve UX for when no address found

### PR DESCRIPTION
**Pull request by samyou-softwire**
[Link to Jira ticket](https://beisdigital.atlassian.net/browse/PC-1463)

# Description

adds routing logic to send directly to address manual if no addresses are found

also adds extra logic on the address page to check whether it will find any addresses. this does mean that each flow that uses OS API will send twice the number of requests. if this is too undesirable then can look into caching the result here

# Checklist

- [x] I have made any necessary updates to the documentation
- [x] I have checked there are no unnecessary IDE warnings in this PR
- [x] I have checked there are no unintentional line ending changes
- [x] I have added tests where applicable
- [x] If I have made any changes to the code, I have used the IDE auto-formatter on it
- [x] If I have made any changes to Python files (even if not changing any content strings), I have run `make extract-translations`
- [x] If I have made any changes to website flow, I have updated the [Flow Miro Board](https://miro.com/app/board/uXjVK4AgQ3E=/)
- [x] If I have made any changes to website flow, I have checked forward and back behaviour is still consistent

# Screenshots

image
image

### CODE REVIEW

File: tests/test_frontdoor.py [hunk 1]
```diff
@@ -2032,41 +2031,6 @@ def test_on_epc_select_page_enter_manually_can_be_selected():
     assert page.has_one('h1:contains("What is the property\'s address?")')
 
 
-@unittest.mock.patch("help_to_heat.frontdoor.interface.EPCApi", MockNotFoundEPCApi)
```
Comment (samyou-softwire) [inline tests/test_frontdoor.py:2035]: this test is no longer meaningful; the user no longer needs to click the select manual link

File: help_to_heat/frontdoor/views.py [hunk 2]
```diff
@@ -635,11 +648,14 @@ def save_post_data(self, data, session_id, page_name):
...
                 else:
                     data[address_choice_field] = address_choice_field_epc_api_fail
+                    check_address_has_results_in_address_api()
             else:
                 data[address_choice_field] = address_choice_field_write_address
+                check_address_has_results_in_address_api()
         except Exception as e:  # noqa: B902
             logger.exception(f"An error occurred: {e}")
             data[address_choice_field] = address_choice_field_epc_api_fail
+            check_address_has_results_in_address_api()
```
Comment (jdgage) [inline help_to_heat/frontdoor/views.py:658]: Doing this feels brittle. I think I'd rather have one call after the try-catch. I don't follow why we don't call it in the one branch - why's that?

**Review by jdgage — CHANGES_REQUESTED** [review]
I'm puzzled by the whole approach: we call the API to check the number of results, then set a flag in the session, then immediately read from that flag again to determine the next page - have I got that right? Why do we need that extra machinery? Can we not pass values through to navigation without saving them to the session?

File: help_to_heat/frontdoor/views.py (hunk 2 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/views.py:658]: in some cases we don't care for what the OS API response is, the only case being if the epc api returns results. could rewrite the logic to include just one call by checking afterwards

File: help_to_heat/frontdoor/views.py [hunk 3]
```diff
@@ -641,6 +649,16 @@ def save_post_data(self, data, session_id, page_name):
             logger.exception(f"An error occurred: {e}")
             data[address_choice_field] = address_choice_field_epc_api_fail
 
+        will_use_os_api = (
+            data[address_choice_field] == address_choice_field_epc_api_fail
+            or country == country_field_scotland
+            and data[address_choice_field] == address_choice_field_write_address
+        )
```
Comment (jdgage) [inline help_to_heat/frontdoor/views.py:656]: Are we now re-implementing logic from above (which could be equally brittle)? This is why I'd consider returning early in the case(s) where we don't want to do it.

**Review by jdgage — CHANGES_REQUESTED** [review]
Another general approach point: are we now making an API call to check how many results, then throwing it away and making the exact same API call on the next page? That feels silly - it's worth interrogating why we can't just fetch the list once.

**Comment by jdgage** [comment]
This looks better: still refer back to the [first comment](https://github.com/UKGovernmentBEIS/help-to-heat-GBIS/pull/410#pullrequestreview-2482589893): why do we need this intermediate field that we write to and then immediately read from?

**Comment by samyou-softwire** [comment]
> I'm puzzled by the whole approach: we call the API to check the number of results, then set a flag in the session, then immediately read from that flag again to determine the next page - have I got that right? Why do we need that extra machinery? Can we not pass values through to navigation without saving them to the session?

in general this is our mechanism for sending information to navigation, saving to session makes the routing reproducible. it's not too different from saving the answer to session and immediately querying that

**Comment by samyou-softwire** [comment]
> > in general this is our mechanism for sending information to navigation, saving to session makes the routing reproducible. it's not too different from saving the answer to session and immediately querying that
> 
> Sure. I think the reason it feels weird is that we've got these 'navigation-only' fields all together with actual user input fields? Can we make the distinction clearer?

have renamed these as 'journey fields' to make the distinction clearer

**Review by jdgage — APPROVED** [review]

**Comment by jdgage** [comment]
Some comments about comments, otherwise all fine

File: help_to_heat/frontdoor/consts.py [hunk 4]
```diff
@@ -149,7 +149,9 @@
 address_choice_field_enter_manually = "enter manually"
 address_building_name_or_number_field = "building_name_or_number"
 address_postcode_field = "postcode"
-address_all_address_and_lmk_details_field = "address_and_lmk_details"
+# PC-1463: removed lmk from field name as it stores info from both flows (lmk & uprn)
```
[the remark this answers is not in the document]
Comment (samyou-softwire) [reply help_to_heat/frontdoor/consts.py:148]: changing the value would mean the key used in the users' session would change and we'd need to consider not breaking existing sessions. the constant name can change for free. I think it's fine for a small semantic change here, though if we'd rather change the value too can look into that

File: help_to_heat/frontdoor/consts.py [hunk 5]
```diff
@@ -149,7 +149,9 @@
...
+address_all_address_and_details_field = "address_and_lmk_details"
+address_no_results_field = "no_results"  # yes/no options
```
[the remark this answers is not in the document]
Comment (samyou-softwire) [reply help_to_heat/frontdoor/consts.py]: #yes/no options just to mark that that field takes a yes or a no. can go and update the rest of such fields to the same effect

File: help_to_heat/frontdoor/consts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdgage) [reply help_to_heat/frontdoor/consts.py:148]: Yes, that's fine, but the comment should say this

[the remark this answers is not in the document]
Comment (samyou-softwire) [reply help_to_heat/frontdoor/consts.py]: have updated this

**Comment by jdgage** [comment]
This is good to go

**Pull request merged**